### ISSUE TITLE
NcHostAgent gets no connections

### ISSUE THREAD

**Issue by markus-lassfolk** [Microsoft/SDN#102]
We've successfully deployed the Network Controller via VMM and added it as a Network Service. 

As far as we can see, every "checkbox" is green. The debug commands shows things are working as expected. But we can't get the Hosts to get any config from the NC. 

As all the NC nodes are on the same subnet, we are using a Cert with the IP Address. And we are using a Cert generated from our CA and not a self-signed cert. 

From one of the Network Controllers:  
```
PS C:\Windows\system32> dir Cert:\LocalMachine\My
dir Cert:\LocalMachine\Root

   PSParentPath: Microsoft.PowerShell.Security\Certificate::LocalMachine\My

Thumbprint                                Subject                                                                                                                                                                 
----------                                -------                                                                                                                                                                 
3D551E6BB84E78E91E9C13C4211B8EA6074295B2  CN=172.16.70.201                                                                                                                                                        


   PSParentPath: Microsoft.PowerShell.Security\Certificate::LocalMachine\Root

Thumbprint                                Subject                                                                                                                                                                 
----------                                -------                                                                                                                                                                 
EC6DE909E6AAF6DDBEBF0C309842E574D7E75A1E  CN=FAHOST12.fabric.customer.com                                                                                                                                     
D60669F654F42A8B75CA2D999E105165276C1815  CN=FAHOST10.fabric.customer.com                                                                                                                                     
CDD4EEAE6000AC7F40C3802C171E30148030C072  CN=Microsoft Root Certificate Authority, DC=microsoft, DC=com                                                                                                           
BE36A4562FB2EE05DBB3D32323ADF445084ED656  CN=Thawte Timestamping CA, OU=Thawte Certification, O=Thawte, L=Durbanville, S=Western Cape, C=ZA                                                                       
A43489159A520F0D93D032CCAF37E7FE20A8B419  CN=Microsoft Root Authority, OU=Microsoft Corporation, OU=Copyright (c) 1997 Microsoft Corp.                                                                            
92B46C76E13054E104F230517E6E504D43AB10B5  CN=Symantec Enterprise Mobile Root for Microsoft, O=Symantec Corporation, C=US                                                                                          
8F43288AD272F3103B6FB1428485EA3014C0BCFE  CN=Microsoft Root Certificate Authority 2011, O=Microsoft Corporation, L=Redmond, S=Washington, C=US                                                                    
7F88CD7223F3C813818C994614A89C99FA3B5247  CN=Microsoft Authenticode(tm) Root Authority, O=MSFT, C=US                                                                                                              
3B1EFD3A66EA28B16697394703A72CA340A05BD5  CN=Microsoft Root Certificate Authority 2010, O=Microsoft Corporation, L=Redmond, S=Washington, C=US                                                                    
35D7A00203C67F1C3F2AB8077993236FDC5B2094  CN=FAHOST11.fabric.customer.com                                                                                                                                     
245C97DF7514E7CF2DF8BE72AE957B9E04741E85  OU=Copyright (c) 1997 Microsoft Corp., OU=Microsoft Time Stamping Service Root, OU=Microsoft Corporation, O=Microsoft Trust Network                                     
18F7C1FCC3090203FD5BAA2F861A754976C8DD25  OU="NO LIABILITY ACCEPTED, (c)97 VeriSign, Inc.", OU=VeriSign Time Stamping Service Root, OU="VeriSign, Inc.", O=VeriSign Trust Network                                 
06B27F5DF10DA174943D521FCB9EE4447E0167FD  CN=FAHOST13.fabric.customer.com                                                                                                                                     
D4DE20D05E66FC53FE1A50882C78DB2852CAE474  CN=Baltimore CyberTrust Root, OU=CyberTrust, O=Baltimore, C=IE                                                                                                          
742C3192E607E424EB4549542BE1BBC53E6174E2  OU=Class 3 Public Primary Certification Authority, O="VeriSign, Inc.", C=US                                                                                             
4EB6D578499B1CCF5F581EAD56BE3D9B6744A5E5  CN=VeriSign Class 3 Public Primary Certification Authority - G5, OU="(c) 2006 VeriSign, Inc. - For authorized use only", OU=VeriSign Trust Network, O="VeriSign, Inc....
02FAF3E291435468607857694DF5E45B68851868  CN=AddTrust External CA Root, OU=AddTrust External TTP Network, O=AddTrust AB, C=SE                                                                                     
A9DF380D933E82934D00440C005EBF35B6A6E7C3  CN=Fabric-root-CA, DC=fabric, DC=customer, DC=com                                                                                                                   
556651B1EDAFB811EBB0E1847F412041064C69FE  CN=Fabric-root-CA, DC=fabric, DC=customer, DC=com                                                                                                                   
```

And from one of the hyper-v Hosts; 
```
PS C:\windows\system32> Debug-NetworkControllerConfigurationState -NetworkController 172.16.70.201 -Credential $creds 
Fetching ResourceType:     accessControlLists
Fetching ResourceType:     servers
Fetching ResourceType:     virtualNetworks
Fetching ResourceType:     networkInterfaces
Fetching ResourceType:     virtualGateways
Fetching ResourceType:     loadbalancerMuxes
Fetching ResourceType:     Gateways

PS C:\windows\system32> get-ItemProperty "hklm:\system\currentcontrolset\services\nchostagent\parameters" -Name HostId |fl HostId
HostId : 2c6263c6-3cbc-4534-b01d-864c77ddcc61

PS C:\windows\system32> Get-NetworkControllerServer -ConnectionUri https://172.16.70.201 -Credential $creds |where { $_.InstanceId -eq "2c6263c6-3cbc-4534-b01d-864c77ddcc61"}

Tags             : 
ResourceRef      : /servers/00000000-0000-0000-0000-0cc47a6ee8f0
InstanceId       : 2c6263c6-3cbc-4534-b01d-864c77ddcc61
Etag             : W/"fe8d07f8-ecee-4107-bede-4bd4caf9db99"
ResourceMetadata : Microsoft.Windows.NetworkController.ResourceMetadata
ResourceId       : 00000000-0000-0000-0000-0cc47a6ee8f0
Properties       : Microsoft.Windows.NetworkController.ServerProperties

PS C:\windows\system32> dir cert:\\localmachine\my
dir cert:\\localmachine\root


   PSParentPath: Microsoft.PowerShell.Security\Certificate::localmachine\my

Thumbprint                                Subject                                                                                                                                          
----------                                -------                                                                                                                                          
96EEAC09C373BD81DD97CCC6B52F526E0E9AD636  CN=Fabric Remote Console Connect, O=Fabric, C=US                                                                                                 
35D7A00203C67F1C3F2AB8077993236FDC5B2094  CN=FAHOST11.fabric.customer.com                                                                                                              
08EC276400F319B4D669650872090AC40135B15D  CN=Microsoft Remote Attestation Service                                                                                                          


   PSParentPath: Microsoft.PowerShell.Security\Certificate::localmachine\root

Thumbprint                                Subject                                                                                                                                          
----------                                -------                                                                                                                                          
EC777213825DACD34033E220845A0ACC263B3022  CN=FAHOST11.fabric.customer.com                                                                                                              
DCDDF2E4BE0DDCC34470BCE10A5BD7F07AA0B528  CN=nccluster.fabric.customer.com                                                                                                             
CDD4EEAE6000AC7F40C3802C171E30148030C072  CN=Microsoft Root Certificate Authority, DC=microsoft, DC=com                                                                                    
BE36A4562FB2EE05DBB3D32323ADF445084ED656  CN=Thawte Timestamping CA, OU=Thawte Certification, O=Thawte, L=Durbanville, S=Western Cape, C=ZA                                                
A43489159A520F0D93D032CCAF37E7FE20A8B419  CN=Microsoft Root Authority, OU=Microsoft Corporation, OU=Copyright (c) 1997 Microsoft Corp.                                                     
92B46C76E13054E104F230517E6E504D43AB10B5  CN=Symantec Enterprise Mobile Root for Microsoft, O=Symantec Corporation, C=US                                                                   
8F43288AD272F3103B6FB1428485EA3014C0BCFE  CN=Microsoft Root Certificate Authority 2011, O=Microsoft Corporation, L=Redmond, S=Washington, C=US                                             
7F88CD7223F3C813818C994614A89C99FA3B5247  CN=Microsoft Authenticode(tm) Root Authority, O=MSFT, C=US                                                                                       
3B1EFD3A66EA28B16697394703A72CA340A05BD5  CN=Microsoft Root Certificate Authority 2010, O=Microsoft Corporation, L=Redmond, S=Washington, C=US                                             
245C97DF7514E7CF2DF8BE72AE957B9E04741E85  OU=Copyright (c) 1997 Microsoft Corp., OU=Microsoft Time Stamping Service Root, OU=Microsoft Corporation, O=Microsoft Trust Network              
18F7C1FCC3090203FD5BAA2F861A754976C8DD25  OU="NO LIABILITY ACCEPTED, (c)97 VeriSign, Inc.", OU=VeriSign Time Stamping Service Root, OU="VeriSign, Inc.", O=VeriSign Trust Network          
16E39B086247CF5F25DFF22A3106EBEF88ED2BEC  CN=ClusInfraCert_20170519_135647013                                                                                                              
D4DE20D05E66FC53FE1A50882C78DB2852CAE474  CN=Baltimore CyberTrust Root, OU=CyberTrust, O=Baltimore, C=IE                                                                                   
742C3192E607E424EB4549542BE1BBC53E6174E2  OU=Class 3 Public Primary Certification Authority, O="VeriSign, Inc.", C=US                                                                      
4EB6D578499B1CCF5F581EAD56BE3D9B6744A5E5  CN=VeriSign Class 3 Public Primary Certification Authority - G5, OU="(c) 2006 VeriSign, Inc. - For authorized use only", OU=VeriSign Trust Net...
3679CA35668772304D30A5FB873B0FA77BB70D54  CN=VeriSign Universal Root Certification Authority, OU="(c) 2008 VeriSign, Inc. - For authorized use only", OU=VeriSign Trust Network, O="Veri...
02FAF3E291435468607857694DF5E45B68851868  CN=AddTrust External CA Root, OU=AddTrust External TTP Network, O=AddTrust AB, C=SE                                                              
A9DF380D933E82934D00440C005EBF35B6A6E7C3  CN=Fabric-root-CA, DC=fabric, DC=customer, DC=com                                                                                            
556651B1EDAFB811EBB0E1847F412041064C69FE  CN=Fabric-root-CA, DC=fabric, DC=customer, DC=com                                                                                            

```

```
PS C:\windows\system32>  netstat -anp tcp |findstr 6640
  TCP    0.0.0.0:6640           0.0.0.0:0              LISTENING
  TCP    172.16.70.130:61098    172.16.70.201:6640     TIME_WAIT
```

The only thing I've been able to find regarding this is; 
> If multiple certificates have the same subject name on the Hyper-V host, the Network Controller Host Agent will randomly choose one to present to the Network Controller. 
But as you can see, there is just one certificate in \my. 

I can also see a lot of Audit Failures in the Security Log on the Network Controller, but can't tell if that's related or "expected". 
![image](https://cloud.githubusercontent.com/assets/3661143/26315335/26faae88-3f11-11e7-827b-6d47aee1baea.png)
I suspect it could be certificate related? Kerberos Constraint Delegation needed? 

Anyone who's got any ideas on how to proceed? 
Is there any way to enable debug logging for the NCHostAgent into EventViewer->NetworkController-NCHostAgent-Admin log? 

Happy to provide any needed information to get to the bottom of this as we've probably re-deployed the Network Controller more than 15 times by now and are 1 week behind schedule.

**Comment by micros01** [comment]
We tried using the IP in the certificate name instead of a DNS-name but didn't get that working, not sure though if using IP was the main reason for things not working at that time, we have redeployed close to 50 times as of now :-)
 
I know MS says you should use IP when on the same subnet but we "ignored" that and ended up using a selfsigned certificate with a "real" DNS name and everything is now working and has been working fine for the last 2 months.

I did check our Security log on the NC's but we do not have any "Audit failures" so that's probably not "expected".

We have however had a couple of issues along the way(to say the least), one thing we did have problems with was at least related to certificates. 

What happened was that we decided to update the BIOS on the hostservers (Fujitsu) before going live, this was done through a windows executable. After a necessary reboot of the hosts we started getting issues with Network policies not working, connection between NC and hosts failed. 

This turned out to be a certificate that was placed in "Trusted Root" by the Bios updater, problem was that "Issued To" and "Issued By" values didn't match, therefore it shouldn't be in "Trusted Root" but in "Intermediate". After disabling/removing that certificate everything started working again..

Not sure if this will help but could be worth a look.

/Michael R
 Make IT

**Comment by markus-lassfolk** [comment]
Thanks Michael. We also initially ignored the "IP" part and used a FQDN for all the first deployments. But as it wasn't working we decided to give the "IP-Cert" a go.  It would be interesting to know why it's a "must use" when the same subnet and why Microsoft does not think a FQDN should always be the preferable way so you are not locked in a subnet-config. 

I'll check the root certificates. Thank you!

**Comment by micros01** [comment]
Good question, if I remember correctly the initial documentation didn't state that it was a must to use IP if on the same subnet. According to the VMM/SDN team the "failover" should be a bit quicker when using IP instead of DNS name so that might be the reason for recommendation.
There is one more thing to think about when using DNS name though, that is to make sure the NC-VMs have permissions to update the DNS record for the clustername, described here: https://docs.microsoft.com/en-us/windows-server/networking/sdn/plan/installation-and-preparation-requirements-for-deploying-network-controller

**Comment by markus-lassfolk** [comment]
I got some answers last night from the SDN team and managed to get it all working. 

My main issue was that the CA Template used for creasing the certificate only included Server Authentication (EKU: 1.3.6.1.5.5.7.3.1). The new documentation here shows it should also include Client Auth [https://docs.microsoft.com/en-us/windows-server/networking/sdn/security/sdn-manage-certs](url). 

With that changed, everything worked a lot better. 

They also confirmed that you can always use FQDN just as Michael R mentioned above with those benefits of using the IP, while FQDN gives more flexibility.

**Comment by markus-lassfolk** [comment]
Problem closed and solved.  Thanks!
